File: src/main/java/com/robohorse/robopojogenerator/generator/postprocessors/KotlinDataClassPostProcessor.java
package com.robohorse.robopojogenerator.generator.postprocessors;

import com.robohorse.robopojogenerator.generator.common.ClassDecorator;
import com.robohorse.robopojogenerator.generator.consts.AnnotationItem;
import com.robohorse.robopojogenerator.generator.consts.ClassTemplate;
import com.robohorse.robopojogenerator.generator.consts.Imports;
import com.robohorse.robopojogenerator.generator.consts.annotations.KotlinAnnotations;
import com.robohorse.robopojogenerator.models.ClassItemModel;
import com.robohorse.robopojogenerator.models.FieldModel;
import com.robohorse.robopojogenerator.models.GenerationModel;

import javax.inject.Inject;
import java.util.Map;
import java.util.Set;

/**
 * This is the KotlinDataClassPostProcessor class
 * Please put some info here.
 *
 * @author Wafer Li
 * @since 16/10/24 00:30
 */
public class KotlinDataClassPostProcessor extends AbsPostProcessor {
    @Inject
    public KotlinDataClassPostProcessor() {
    }

    @Override
    protected StringBuilder proceedClassImports(ClassItemModel classItemModel) {
        final Set<String> imports = classItemModel.getClassImports();
        imports.remove(Imports.LIST);
        final StringBuilder importsBuilder = new StringBuilder();
        for (String importItem : imports) {
            importsBuilder.append(importItem.replace(";", ""));
            importsBuilder.append(ClassTemplate.NEW_LINE);
        }
        return importsBuilder;
    }

    @Override
    public String proceedClassBody(ClassItemModel classItemModel, GenerationModel generationModel) {
        final StringBuilder classBodyBuilder = new StringBuilder();
        final Map<String, ClassDecorator> classFields = classItemModel.getClassFields();
        for (String objectName : classFields.keySet()) {
            classBodyBuilder.append(classTemplateHelper.createKotlinDataClassField(
                    new FieldModel.Builder()
                            .setFieldName(objectName)
                            .setClassType(classFields.get(objectName).getKotlinItem())
                            .setAnnotation(classItemModel.getAnnotation())
                            .setFieldNameFormatted(generateHelper.formatClassField(objectName))
                            .build()
            ));
        }
        generateHelper.updateClassModel(classBodyBuilder);
        return classBodyBuilder.toString();
    }

    @Override
    protected String createClassItemText(String packagePath, String imports, String classTemplate) {
        return classTemplateHelper
                .createClassItemWithoutSemicolon(
                        packagePath,
                        imports,
                        classTemplate);
    }

    @Override
    protected void applyAnnotations(AnnotationItem item, ClassItemModel classItemModel) {
        switch (item) {
            case GSON: {
                generateHelper.setAnnotations(classItemModel,
                        KotlinAnnotations.GSON.CLASS_ANNOTATION,
                        KotlinAnnotations.GSON.ANNOTATION,
                        Imports.GSON.IMPORTS);
                break;
            }
            case LOGAN_SQUARE: {
                generateHelper.setAnnotations(classItemModel,
                        KotlinAnnotations.LOGAN_SQUARE.CLASS_ANNOTATION,
                        KotlinAnnotations.LOGAN_SQUARE.ANNOTATION,
                        Imports.LOGAN_SQUARE.IMPORTS);
                break;
            }
            case JACKSON: {
                generateHelper.setAnnotations(classItemModel,
                        KotlinAnnotations.JACKSON.CLASS_ANNOTATION,
                        KotlinAnnotations.JACKSON.ANNOTATION,
                        Imports.JACKSON.IMPORTS);
                break;
            }
        }
    }

    @Override
    public String createClassTemplate(ClassItemModel classItemModel, String classBody) {
        return classTemplateHelper.createClassBodyKotlinDataClass(classItemModel, classBody);
    }
}

File: generator/src/main/kotlin/com/robohorse/robopojogenerator/postrocessing/common/KotlinDataClassPostProcessor.kt
package com.robohorse.robopojogenerator.postrocessing.common

import com.robohorse.robopojogenerator.models.FieldModel
import com.robohorse.robopojogenerator.models.FrameworkVW.*
import com.robohorse.robopojogenerator.models.GenerationModel
import com.robohorse.robopojogenerator.postrocessing.BasePostProcessor
import com.robohorse.robopojogenerator.postrocessing.utils.MoshiAnnotationsProcessor
import com.robohorse.robopojogenerator.properties.ClassItem
import com.robohorse.robopojogenerator.properties.annotations.KotlinAnnotations
import com.robohorse.robopojogenerator.properties.templates.ClassTemplate
import com.robohorse.robopojogenerator.properties.templates.ImportsTemplate
import com.robohorse.robopojogenerator.properties.templates.PARCELABLE_ANDROID
import com.robohorse.robopojogenerator.properties.templates.PARCELIZE_KOTLINX
import com.robohorse.robopojogenerator.utils.ClassGenerateHelper
import com.robohorse.robopojogenerator.utils.ClassTemplateHelper

internal class KotlinDataClassPostProcessor(
    generateHelper: ClassGenerateHelper,
    classTemplateHelper: ClassTemplateHelper,
    private val moshiAnnotationsProcessor: MoshiAnnotationsProcessor
) : BasePostProcessor(generateHelper, classTemplateHelper) {

    override fun proceedClassImports(
        imports: HashSet<String>,
        generationModel: GenerationModel
    ): StringBuilder {
        imports.remove(ImportsTemplate.LIST)
        if (generationModel.useKotlinParcelable) {
            imports.add(PARCELABLE_ANDROID)
            imports.add(PARCELIZE_KOTLINX)
        }
        val importsBuilder = StringBuilder()
        for (importItem in imports) {
            importsBuilder.append(importItem.replace(";", ""))
            importsBuilder.append(ClassTemplate.NEW_LINE)
        }
        return importsBuilder
    }

    override fun proceedClassBody(
        classItem: ClassItem,
        generationModel: GenerationModel
    ): String {
        val classBodyBuilder = StringBuilder()
        val classFields = classItem.classFields
        for (objectName in classFields.keys) {
            classBodyBuilder.append(
                classTemplateHelper.createKotlinDataClassField(
                    generationModel,
                    FieldModel(
                        classType = classFields[objectName]?.getKotlinItem(),
                        annotation = classItem.annotation,
                        fieldName = objectName,
                        fieldNameFormatted = generateHelper.formatClassField(objectName)
                    )
                )
            )
        }
        generateHelper.updateClassModel(classBodyBuilder)
        return classBodyBuilder.toString()
    }

    override fun createClassItemText(
        packagePath: String?,
        imports: String?,
        classTemplate: String?
    ) = classTemplateHelper
        .createClassItemWithoutSemicolon(
            packagePath,
            imports,
            classTemplate
        )

    override fun applyAnnotations(
        generationModel: GenerationModel,
        classItem: ClassItem
    ) = when (generationModel.annotationEnum) {
        is Gson -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.GSON.classAnnotation,
                KotlinAnnotations.GSON.annotation,
                ImportsTemplate.GSON.imports
            )
        }

        is KotlinX -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.KOTLIN_X.classAnnotation,
                KotlinAnnotations.KOTLIN_X.annotation,
                ImportsTemplate.KOTLIN_X.imports
            )
        }

        is LoganSquare -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.LOGAN_SQUARE.classAnnotation,
                KotlinAnnotations.LOGAN_SQUARE.annotation,
                ImportsTemplate.LOGAN_SQUARE.imports
            )
        }

        is Jackson -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.JACKSON.classAnnotation,
                KotlinAnnotations.JACKSON.annotation,
                ImportsTemplate.JACKSON.imports
            )
        }

        is FastJson -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.FAST_JSON.classAnnotation,
                KotlinAnnotations.FAST_JSON.annotation,
                ImportsTemplate.FAST_JSON.imports
            )
        }

        is Jakatra -> {
            generateHelper.setAnnotations(
                classItem,
                KotlinAnnotations.JAKATRA.classAnnotation,
                KotlinAnnotations.JAKATRA.annotation,
                ImportsTemplate.JAKATRA.imports
            )
        }

        is Moshi -> moshiAnnotationsProcessor.applyAnnotations(generationModel, classItem)

        else -> {
            // NO OP
        }
    }

    override fun createClassTemplate(
        classItem: ClassItem,
        classBody: String?,
        generationModel: GenerationModel
    ) = classTemplateHelper.createClassBodyKotlinDataClass(
        classItem,
        classBody,
        generationModel
    )
}
